npeps/npep-126-egress-traffic-control.md

```diff
@@ -76,13 +76,184 @@ selected cluster workloads to k8s-apiservers for securing the server.
+
+```
+// AdminNetworkPolicyEgressPeer defines a peer to allow traffic to.
+// Exactly one of the selector pointers must be set for a given peer. If a
+// consumer observes none of its fields are set, they must assume an unknown
+// option has been specified and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+	// Nodes defines a way to select a set of nodes in
+	// in the cluster. This field follows standard label selector
+	// semantics; if present but empty, it selects all Nodes.
+	//
+	// Support: Core
+	//
+	// +optional
+	Nodes *metav1.LabelSelector `json:"nodes,omitempty"`
+}
+```
+
+Note that `AdminNetworkPolicyPeer` will be changed to
+`AdminNetworkPolicyEgressPeer` and `AdminNetworkPolicyIngressPeer` since ingress and
+egress peers have started to diverge at this point and it is easy to
+maintain it with two sets of peer definitions.
+This ensures nodes can be referred to only as "egress peers".
+
+Example: Admin wants to deny egress traffic from tenants who don't have
+`restricted`, `confidential` or `internal` level security clearance
+to control-plane nodes at 443 and 6443 ports in the cluster
+
+```
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: AdminNetworkPolicy
+metadata:
+  name: node-as-egress-peers
+spec:
+  priority: 55
+  subject:
+    namespaces:
+      matchExpressions:
+        - {key: security, operator: notIn, values: [restricted, confidential, internal]}
+  egress:
+  - name: "deny-all-egress-to-kapi-server"
+    action: "Deny"
+    to:
+    - nodes:
+        matchLabels:
+          node-role.kubernetes.io/control-plane:
+    ports:
+      - portNumber:
+          protocol: TCP
+          port: 443
+      - portNumber:
+          protocol: TCP
+          port: 6443
+```
+
+### Implementing egress traffic control towards external destinations
+
+This NPEP proposes to add a new type of `AdminNetworkPolicyEgressPeer` called `Networks`
+to be able to explicitly select external destination CIDRs outside the cluster.
+This peer type will not be supported in `AdminNetworkPolicyIngressPeer`.
+
+```
+// AdminNetworkPolicyEgressPeer defines a peer to allow traffic to.
+// Exactly one of the selector pointers must be set for a given peer. If a
+// consumer observes none of its fields are set, they must assume an unknown
+// option has been specified and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+	// Networks defines a way to select IPs and CIDR blocks that represent
+  // entities that live outside the cluster as a peer.
+	// It is the list of NetworkCIDR (both v4 & v6) that can be used to define
+	// external destinations.
+  // You could use this to select internal cluster networks like podCIDR blocks, but it is
+  // recommended to use namespaces and pods peers instead in those cases.
```

Comment:
tab/space issues?

"NetworkCIDR" does not appear anywhere else in the NPEP?

"IPs and CIDR blocks" implies you could use a syntactically-just-an-IP value, which you can't

and don't say "outside the cluster" or "external"...

```suggestion
	// Networks defines a way to select peers via CIDR blocks. This is
	// intended for representing entities that live outside the cluster,
	// which can't be selected by pods and namespaces peers, but note
	// that cluster-internal traffic will be checked against the rule as
	// well, so if you Allow or Deny traffic to `"0.0.0.0/0"`, that will allow
	// or deny all IPv4 pod-to-pod traffic as well. If you don't want that,
	// add a rule that Passes all pod traffic before the Networks rule.
```

Reply:
unsure what's with the space/tab thing. It looked ok in my editor:
![Screenshot from 2024-01-22 10-18-46](https://github.com/kubernetes-sigs/network-policy-api/assets/4757069/d3730b40-b99c-462b-a144-4a2833910e25)
will fix it...